src/basic/json.h

```diff
@@ -0,0 +1,272 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#pragma once
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "macro.h"
+#include "string-util.h"
+#include "util.h"
+
+/*
+  In case you wonder why we have our own JSON implementation, here are a couple of reasons why this implementation has
+  benefits over various other implementatins:
```

Comment:
I'm not going to say it's a bold move to implement a json parser :-), but if it's going to be implemented, maybe it would be helpful to start fuzzing it right away. I ran a fuzzer  compiled with UBSan for about ten seconds and it crashed after receiving relatively normal input like `-34444444444541444444` and `"0"` with
```sh
../../src/systemd/src/basic/json.c:1845:48: runtime error: signed integer overflow: 10 * -3444444444454144444 cannot be represented in type 'long'
    #0 0x7fe6beefb805 in json_parse_number /work/build/../../src/systemd/src/basic/json.c:1845:48
    #1 0x7fe6beefa5ca in json_tokenize /work/build/../../src/systemd/src/basic/json.c:2009:29
```
and
```
../../src/systemd/src/basic/json.c:350:9: runtime error: index 1 out of bounds for type 'char [0]'
    #0 0x7fab2d4c2042 in json_variant_new_stringn /work/build/../../src/systemd/src/basic/json.c:350:22
    #1 0x7fab2d4c8690 in json_parse_internal /work/build/../../src/systemd/src/basic/json.c:2321:29
    #2 0x7fab2d4c781a in json_parse /work/build/../../src/systemd/src/basic/json.c:2477:16
```
There were also a couple warnings from `clang`, the first of which seems legit:
```
../../src/systemd/src/basic/json.c:2486:32: warning: unused variable 'opened' [-Wunused-variable]
        _cleanup_fclose_ FILE *opened = NULL;
                               ^
../../src/systemd/src/basic/json.c:2949:13: warning: variable 'source' is used uninitialized whenever 'if' condition is false [-Wsometimes-uninitialized]
        if (variant) {
            ^~~~~~~
../../src/systemd/src/basic/json.c:2955:13: note: uninitialized use occurs here
        if (source && source_line > 0 && source_column > 0)
            ^~~~~~
../../src/systemd/src/basic/json.c:2949:9: note: remove the 'if' if its condition is always true
        if (variant) {
        ^~~~~~~~~~~~~
../../src/systemd/src/basic/json.c:2936:27: note: initialize the variable 'source' to silence this warning
        const char *source;
                          ^
                           = NULL
```

Reply:
Unfortunately, false positives are inevitable, but I don't think they make fuzzers less useful. In this particular case, `__attribute__((no_sanitize("signed-integer-overflow")))` and `__attribute__((no_sanitize("bounds")))` could probably be used to silence UBSan.

Reply:
I figure I could rewrite the overflow check to not trigger ubsan. But for the [0] bounds check I have no idea how I could trick it to not complain. Any idea?

Quite frankly, ubsan should be smart enough to understand that [0] is special, and generally doesn't actually mean "zero sized array", but instead is similar to C99 [], i.e. more akin to "unspecified size"...

Reply:
I've done some research and it turns out that UBSan is smart enough to understand that [0] is special as long as it's the last element of a union within a structure. Placing `char string[0]` at the end of the union will make the warning go away.

Reply:
BTW I've written a fuzzer. It's waiting to be unleashed in https://github.com/systemd/systemd/pull/9782 :-)

Reply:
@evverx hmm, I figure it would make sense to simply pick up that #9782 and add it to this PR, what do you think?

Reply:
@poettering it's fine by me, but it shouldn't be merged into master until https://github.com/google/oss-fuzz/pull/1683 is merged.

Reply:
@poettering https://github.com/google/oss-fuzz/pull/1683 has been merged. I pushed the commit from #9782 on top of your branch.

Reply:
Fedora CI seems to have vanished. Does anybody know how to bring it back or trigger a build manually?

Reply:
I'm not sure what's happened, but the fuzzer is gone. It'd probably be better to leave it util later. I'll reopen #9782 so as not to forget that it isn't done yet.